src/components/CodeBlock.js

```diff
@@ -0,0 +1,113 @@
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import cx from 'classnames';
+import Button from './Button';
+import CodeEditor from './CodeEditor';
+import CodeHighlight from './CodeHighlight';
+import FeatherIcon from './FeatherIcon';
+import MiddleEllipsis from 'react-middle-ellipsis';
+import { LiveError, LivePreview, LiveProvider } from 'react-live';
+import styles from './CodeBlock.module.scss';
+import useClipboard from '../hooks/useClipboard';
+import useFormattedCode from '../hooks/useFormattedCode';
+
+const defaultComponents = {
+  Preview: LivePreview,
+};
+
+const CodeBlock = ({
+  children,
+  components: componentOverrides = {},
```

Comment:
Do we have any instances where we are adding components that _aren't_ the `Preview`? I'm a little unclear about the purpose of this prop.

Reply:
No. `<Preview />` is the only component right now that is customizable. To understand how I got here, let me describe a few of my goals for this refactor and some ways I considered designing this API.

#### Goals

1. We can use this anywhere in the site that needs a code block. 

Regardless of where it is rendered, it should be able to handle all use cases demanded of the code block. This includes: 
* conditionally showing line numbers
* conditionally allowing people to copy the code
* Allowing users to edit the code inline (useful for component docs)
* Show a live preview of the code snippet (also useful for component docs)

2. Allow this component to be used to show live previews in guides

While this is [explicitly disabled](https://github.com/newrelic/developer-website/pull/409/files#diff-b55d88313c0add65aa1df80e51c1f1adR33) currently, its something we can enable when we have some demand for this particular use case.

3. With the upcoming move to a shared theme, we should be able to cut/paste this component into the shared theme library and use it across all sites.

The API is designed to be used everywhere. You'll notice that there is no developer website specific code in here. We want to be able to use this component in the open source site as well, which means we have to try and be as generic as possible.

#### API Design

Goal 3 presents an interesting challenge. Because we want to enable live previews on any code on any site, not just the NR1 SDK in the developer site, we can't use something like the code in `<ReferencePreview />` which uses shadow DOM, loads the NR1 SDK's CSS, etc. On the flip side, because we want to be generic about this, this also presents a problem for our use case of needing style isolation for the NR1 SDK component previews. This means that the live preview **MUST** be customizable. 

Specifically, the NR1 SDK component previews require the following:
* Must be rendered in the shadow DOM to provide style isolation
* Must be able to load an external stylesheet
* Must be able to define inline CSS for additional styling (useful for guides like the `<Grid />` which show boxes)
* Must be able to render other components (used to mount the `ToastManager` when the `Toast` component is used)
* Must be able to define additional styles on the root preview container (useful for examples like the `<Spinner />` which renders absolutely positioned)

There are a couple of APIs I considered to allow this to be customizable with the previous requirements in mind:

1. Allow extra props to be passed to configure the preview component:

```js
<CodeBlock
  preview={true}
  previewExternalStylesheet="https://some.external.url/example.css"
  previewInShadowDOM={true}
  previewStyle={{ position: 'relative' }}
  previewChildren={() => <ToastManager />}
>
  import React from 'react'
  // the rest of the code that should be highlighted by the code block
</CodeBlock>
```

While these may not have been the final names of the props, you can see that the surface area of `<CodeBlock />` now has grown a lot, specifically with a LOT of props dedicated to configuring the preview. This approach also means we need to add new props any time we have new uses cases for the preview that weren't already considered.

2. Use the `.` delimited approach to give back control of the structure to the user of the component

```js
import root from 'react-shadow'

<CodeBlock>
  <root.div>
    <link href='stylesheet.css' />
    <style type='text/css'>.h1 { font-weight: bold; }</style>
    <CodeBlock.Preview />
  </root.div>
  <CodeBlock.Code>
    import React from 'react'
    // the rest of the code that should be highlighted by the code block
  <CodeBlock.Code>
</CodeBlock>
```

While this approach is a bit better because I no longer have tons of `preview` props dedicated to configuring the component preview, I've now lost control of the structure of `<CodeBlock />` and can't enforce it. For example, what if someone uses the component this way?

```js
<CodeBlock>
  <CodeBlock.Code />
  <CodeBlock.Preview />
</CodeBlock>
```

Do we honor this structure and render the preview underneath the code? And what about the status bar that renders the file name and copy button? Are those lost? Should we provide a component for that to make it obvious it's being rendered?

As you can see, while I gain the ability to easily customize the structure of the preview, I've now lost an enforced structure for the entire thing and put the burden back into the hands of the user of the component to always make sure these are in the proper order. I didn't like this unnecessary burden since we want an enforced structure for the code block.

3. Allow a user to provide their own component for the preview

```js
const MyCustomPreview = ({ children, className }) => (
  <root.div>
    <link href="some.css" />
    {children}
  </root.div>
)

<CodeBlock
  preview={true}
  components={{ Preview: MyCustomPreview }}
/>
```

This is popular with libraries like [React Select](https://react-select.com/home) and even the `MDXProvider` for rendering mdx with customized components.

This approach allows us to be able to enforce the proper structure internally to the code block, but also provides the flexibility for the user to customize the structure of the preview component. This means users of the API can design the preview for any and all uses cases that may pop up over time without needing to touch the `<CodeBlock />` to do so. This is the approach I opted for as I felt it provided the greatest flexibility without compromising the enforced structure we want in this component.

#### Final notes

Right now `Preview` is the only component that is allowed to be customized because it is the only one that needs it. I'm not sure it makes sense to provide a customized status bar, or customized code highlighting as of right now. If that use case ever exists, then we have a way to extend this component to allow for those customizations.

Hope that explains the reasoning behind this API design choice and why I chose to go the way I did.